glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+/// to two dimensions rather than the single label provided by the standard labeled type.
+#[derive(Debug)]
+pub struct DualLabeledCounterMetric {
+    keys: Option<Vec<Cow<'static, str>>>,
+    categories: Option<Vec<Cow<'static, str>>>,
+    /// Type of the underlying metric
+    /// We hold on to an instance of it, which is cloned to create new modified instances.
+    counter: CounterMetric,
+
+    /// A map from a unique ID for the dual labeled submetric to a handle of an instantiated
+    /// metric type.
+    dual_label_map: Mutex<HashMap<(String, String), Arc<CounterMetric>>>,
+}
+
+// TODO figure out how to measure the sizeof these metrics
+// impl ::malloc_size_of::MallocSizeOf for DualLabeledCounterMetric {
+//     fn size_of(&self, ops: &mut malloc_size_of::MallocSizeOfOps) -> usize {
+//         let map = self.dual_label_map.lock().unwrap();
+//         let keys: Vec<String> = map.keys().collect();
+//         let categories = map.values().for_each(|category| -> String  { category.keys() });
+
+//         // Copy of `MallocShallowSizeOf` implementation for `HashMap<K, V>` in `wr_malloc_size_of`.
+//         // Note: An instantiated submetric is behind an `Arc`.
+//         // `size_of` should only be called from a single thread to avoid double-counting.
+//         let shallow_size = if ops.has_malloc_enclosing_size_of() {
+//             map.values()
+//                 .next()
+//                 .map_or(0, |v| unsafe { ops.malloc_enclosing_size_of(v) })
+//         } else {
+//             map.capacity()
+//                 * (mem::size_of::<String>() + mem::size_of::<T>() + mem::size_of::<usize>())
+//         };
+
+//         let mut map_size = shallow_size;
+//         for (k, v) in map.iter() {
+//             map_size += k.size_of(ops);
+//             map_size += v.size_of(ops);
+//         }
+
+//         self.labels.size_of(ops) + self.submetric.size_of(ops) + map_size
+//     }
+// }
+
+impl DualLabeledCounterMetric {
+    /// Creates a new dual labeled counter from the given metric instance and optional list of labels.
+    pub fn new(
+        meta: CommonMetricData,
+        keys: Option<Vec<Cow<'static, str>>>,
+        catgories: Option<Vec<Cow<'static, str>>>,
+    ) -> DualLabeledCounterMetric {
+        let submetric = CounterMetric::new(meta);
+        DualLabeledCounterMetric::new_inner(submetric, keys, catgories)
+    }
+
+    fn new_inner(
+        counter: CounterMetric,
+        keys: Option<Vec<Cow<'static, str>>>,
+        categories: Option<Vec<Cow<'static, str>>>,
+    ) -> DualLabeledCounterMetric {
+        let dual_label_map = Default::default();
+        DualLabeledCounterMetric {
+            keys,
+            categories,
+            counter,
+            dual_label_map,
+        }
+    }
+
+    /// Creates a new metric with a specific label.
+    ///
+    /// This is used for dynamic labels where we have to actually validate and correct the
+    /// label later when we have a Glean object.
```

Comment:
This comment is slightly incorrect.
For labeled metrics we split things into two functions, but here both static and dynamic labels are handled in this function.